File: haskell/glue/src/Glue/Lib/Math/Arithmetic/Sub.hs
module Glue.Lib.Math.Arithmetic.Sub where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Subtraction function
-- Mirrors Haskell Glue.Lib.Math.Arithmetic.Sub.sub exactly
sub :: IR Eval
sub = NativeFunc subImpl

-- Subtraction function implementation
-- Mirrors Haskell Glue.Lib.Math.Arithmetic.Sub.subImpl exactly
subImpl :: IR Eval -> Eval (IR Eval)
subImpl left = pure $ NativeFunc (subFrom left)

subFrom :: IR Eval -> IR Eval -> Eval (IR Eval)
subFrom left right = case (left, right) of
    (Integer a, Integer b) -> pure $ Integer (a - b)
    (Integer a, Float b) -> pure $ Float (fromIntegral a - b)
    (Float a, Integer b) -> pure $ Float (a - fromIntegral b)
    (Float a, Float b) -> pure $ Float (a - b)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/arithmetic/sub.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Subtraction function
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Sub.sub exactly
final Ir sub = IrNativeFunc(subImpl);

/// Subtraction function implementation
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Sub.subImpl exactly
Eval<Ir> subImpl(Ir left) {
  return Eval.pure(IrNativeFunc(subFrom(left)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Sub.subFrom exactly
Eval<Ir> Function(Ir) subFrom(Ir left) {
  return (Ir right) {
    return sequenceAll([eval(left), eval(right)]).flatMap((values) {
      return switch (values) {
        [IrInteger(value: final a), IrInteger(value: final b)] => Eval.pure(
          IrInteger(a - b),
        ),
        [IrInteger(value: final a), IrFloat(value: final b)] => Eval.pure(
          IrFloat(a - b),
        ),
        [IrFloat(value: final a), IrInteger(value: final b)] => Eval.pure(
          IrFloat(a - b),
        ),
        [IrFloat(value: final a), IrFloat(value: final b)] => Eval.pure(
          IrFloat(a - b),
        ),
        _ => throwError(wrongArgumentType(['number'])),
      };
    });
  };
}
